provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -538,6 +540,108 @@ func megabytesToKiB(diskMiB uint64) uint64 {
 	return diskMiB * 1024
 }
 
+func (c *Client) getMaxSuportedVersion(ctx context.Context, cr *mo.ComputeResource) (int64, error) {
+	if cr == nil || cr.EnvironmentBrowser == nil {
+		return 0, errors.Errorf("invalid compute resource")
+	}
+	ref := cr.EnvironmentBrowser.Reference()
+	req := types.QueryConfigOption{
+		This: ref,
+	}
+	opt, err := methods.QueryConfigOption(ctx, c.client, &req)
+	if err != nil {
+		return 0, err
+	}
+
+	if opt.Returnval == nil {
+		return 0, fmt.Errorf("failed to get max supported version")
+	}
+
+	parsed, err := c.parseVMXVersion(opt.Returnval.Version)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	return parsed, nil
+}
+
+func (c *Client) getVMHardwareVersion(ctx context.Context, srcVM *object.VirtualMachine) (int64, error) {
+	if srcVM == nil {
+		return 0, errors.Errorf("source VM may not be nil")
+	}
+	var templateVM mo.VirtualMachine
+	err := srcVM.Properties(ctx, srcVM.Reference(), []string{"config.version"}, &templateVM)
+	if err != nil {
+		return 0, err
+	}
+
+	if templateVM.Config == nil {
+		return 0, fmt.Errorf("failed to get VM hardware version")
+	}
+	parsed, err := c.parseVMXVersion(templateVM.Config.Version)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	return parsed, nil
+}
+
+func (c *Client) parseVMXVersion(version string) (int64, error) {
+	fields := strings.Split(version, "-")
+	if len(fields) != 2 {
+		return 0, fmt.Errorf("invalid VMX version: %s", version)
```

Comment:
Can we use `errors.Errorf`? Using the `errors` package will give us a stack trace